// amplify-python-function-runtime-provider/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.functionRuntimeContributorFactory = void 0;
const buildUtils_1 = require("./util/buildUtils");
const packageUtils_1 = require("./util/packageUtils");
const invokeUtil_1 = require("./util/invokeUtil");
const depUtils_1 = require("./util/depUtils");
const path_1 = __importDefault(require("path"));
exports.functionRuntimeContributorFactory = context => {
    return {
        contribute: request => {
            const selection = request.selection;
            if (selection !== 'python') {
                return Promise.reject(new Error(`Unknown selection ${selection}`));
            }
            return Promise.resolve({
                runtime: {
                    name: 'Python',
                    value: 'python',
                    cloudTemplateValue: 'python3.8',
                    defaultHandler: 'index.handler',
                    layerExecutablePath: path_1.default.join('python', 'lib', 'python3.8', 'site-packages'),
                },
            });
        },
        checkDependencies: depUtils_1.checkDeps,
        package: request => packageUtils_1.pythonPackage(context, request),
        build: buildUtils_1.pythonBuild,
        invoke: request => buildUtils_1.pythonBuild(request).then(() => invokeUtil_1.pythonInvoke(context, request)),
    };
};
