// @aws-amplify/cli/lib/extensions/amplify-helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.amplifyCLIConstants = void 0;
exports.amplifyCLIConstants = {
    AmplifyCLIDirName: 'amplify',
    DotAmplifyDirName: '.amplify',
    DotConfigamplifyCLISubDirName: '.config',
    BackendamplifyCLISubDirName: 'backend',
    CurrentCloudBackendamplifyCLISubDirName: '#current-cloud-backend',
    ProjectConfigFileName: 'project-config.json',
    amplifyMetaFileName: 'amplify-meta.json',
    BrandName: 'Amplify Framework',
    CliName: 'amplify',
    DateTimeFormatString: 'YYYY-MM-DD-HH-mm-ss',
    DateTimeFormatStringCompact: 'YYYYMMDDHHmmss',
    DefaultAWSAccessKeyId: '<YOUR_ACCESS_KEY_ID>',
    DefaultAWSSecretAccessKey: '<YOUR_SECRET_ACCESS_KEY>',
    DefaultAWSRegion: 'us-east-1',
    LocalEnvFileName: 'local-env-info.json',
    ProviderInfoFileName: 'team-provider-info.json',
    BackendConfigFileName: 'backend-config.json',
    TagsFileName: 'tags.json',
    PROJECT_CONFIG_VERSION: '3.0',
    BreadcrumbsFileName: 'amplify.state',
};
