// @aws-amplify/cli/lib/extensions/amplify-helpers/editor-selection.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.normalizeEditor = exports.editorSelection = exports.editors = void 0;
const inquirer = __importStar(require("inquirer"));
exports.editors = [
    {
        name: 'Visual Studio Code',
        value: 'vscode',
    },
    {
        name: 'Atom Editor',
        value: 'atom',
    },
    {
        name: 'Sublime Text',
        value: 'sublime',
    },
    {
        name: 'IntelliJ IDEA',
        value: 'intellij',
    },
    {
        name: 'Vim (via Terminal, Mac OS only)',
        value: 'vim',
    },
    {
        name: 'Emacs (via Terminal, Mac OS only)',
        value: 'emacs',
    },
    {
        name: 'None',
        value: 'none',
    },
];
async function editorSelection(defaultEditor) {
    const normalizedDefaultEditor = exports.editors.findIndex(editor => editor.value === defaultEditor) > -1 ? defaultEditor : undefined;
    const editorQuestion = {
        type: 'list',
        name: 'editorSelected',
        message: 'Choose your default editor:',
        default: normalizedDefaultEditor,
        choices: exports.editors,
    };
    const { editorSelected } = await inquirer.prompt(editorQuestion);
    return editorSelected;
}
exports.editorSelection = editorSelection;
function normalizeEditor(editor) {
    if (editor) {
        editor = editor.toLowerCase();
        if (editor === 'idea14ce') {
            editor = 'intellij';
        }
        else if (editor === 'code') {
            editor = 'vscode';
        }
        editor = exports.editors.findIndex(editorEntry => editorEntry.value === editor) > -1 ? editor : undefined;
    }
    return editor;
}
exports.normalizeEditor = normalizeEditor;
