// @aws-amplify/cli/lib/extensions/amplify-helpers/service-select-prompt.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serviceSelectionPrompt = void 0;
const inquirer = __importStar(require("inquirer"));
const get_project_config_1 = require("./get-project-config");
const get_provider_plugins_1 = require("./get-provider-plugins");
const amplify_cli_core_1 = require("amplify-cli-core");
function filterServicesByEnabledProviders(context, enabledProviders, supportedServices) {
    const providerPlugins = get_provider_plugins_1.getProviderPlugins(context);
    const filteredServices = [];
    Object.keys(supportedServices).forEach(service => {
        if (enabledProviders.includes(supportedServices[service].provider)) {
            filteredServices.push({
                service,
                providerPlugin: providerPlugins[supportedServices[service].provider],
                providerName: supportedServices[service].provider,
                alias: supportedServices[service].alias,
            });
        }
    });
    return filteredServices;
}
function serviceQuestionWalkthrough(context, supportedServices, category, customQuestion = null) {
    const options = [];
    for (let i = 0; i < supportedServices.length; ++i) {
        const optionName = supportedServices[i].alias || `${supportedServices[i].providerName}:${supportedServices[i].service}`;
        options.push({
            name: optionName,
            value: {
                provider: supportedServices[i].providerPlugin,
                service: supportedServices[i].service,
                providerName: supportedServices[i].providerName,
            },
        });
    }
    if (options.length === 0) {
        const errMessage = `No services defined by configured providers for category: ${category}`;
        context.print.error(errMessage);
        context.usageData.emitError(new amplify_cli_core_1.ResourceDoesNotExistError(errMessage));
        amplify_cli_core_1.exitOnNextTick(1);
    }
    if (options.length === 1) {
        context.print.info(`Using service: ${options[0].value.service}, provided by: ${options[0].value.providerName}`);
        return new Promise(resolve => {
            resolve(options[0].value);
        });
    }
    const question = [
        {
            name: 'service',
            message: customQuestion || 'Please select from one of the below mentioned services:',
            type: 'list',
            choices: options,
        },
    ];
    return inquirer.prompt(question).then(answer => answer.service);
}
function serviceSelectionPrompt(context, category, supportedServices, customQuestion = null) {
    const { providers } = get_project_config_1.getProjectConfig();
    supportedServices = filterServicesByEnabledProviders(context, providers, supportedServices);
    return serviceQuestionWalkthrough(context, supportedServices, category, customQuestion);
}
exports.serviceSelectionPrompt = serviceSelectionPrompt;
