// amplify-cli-core/lib/errors/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrontendBuildError = exports.InvalidSubCommandError = exports.InvalidEnvironmentNameError = exports.NonEmptyDirectoryError = exports.MissingParametersError = exports.EnvironmentDoesNotExistError = exports.UnknownArgumentError = exports.UnknownResourceTypeError = exports.ResourceCredentialsNotFoundError = exports.ResourceDoesNotExistError = exports.ResourceAlreadyExistsError = exports.NotImplementedError = void 0;
class NotImplementedError extends Error {
}
exports.NotImplementedError = NotImplementedError;
class ResourceAlreadyExistsError extends Error {
}
exports.ResourceAlreadyExistsError = ResourceAlreadyExistsError;
class ResourceDoesNotExistError extends Error {
}
exports.ResourceDoesNotExistError = ResourceDoesNotExistError;
class ResourceCredentialsNotFoundError extends Error {
}
exports.ResourceCredentialsNotFoundError = ResourceCredentialsNotFoundError;
class UnknownResourceTypeError extends Error {
}
exports.UnknownResourceTypeError = UnknownResourceTypeError;
class UnknownArgumentError extends Error {
}
exports.UnknownArgumentError = UnknownArgumentError;
class EnvironmentDoesNotExistError extends Error {
}
exports.EnvironmentDoesNotExistError = EnvironmentDoesNotExistError;
class MissingParametersError extends Error {
}
exports.MissingParametersError = MissingParametersError;
class NonEmptyDirectoryError extends Error {
}
exports.NonEmptyDirectoryError = NonEmptyDirectoryError;
class InvalidEnvironmentNameError extends Error {
}
exports.InvalidEnvironmentNameError = InvalidEnvironmentNameError;
class InvalidSubCommandError extends Error {
}
exports.InvalidSubCommandError = InvalidSubCommandError;
class FrontendBuildError extends Error {
}
exports.FrontendBuildError = FrontendBuildError;
